### DISCUSSION TITLE
Binding for Telegram's tdlib. Symbol not found: _typeinfo for td::td_api::logStreamDefault

### DISCUSSION THREAD

**Discussion by MarshalX** [pybind/pybind11#3171] [Q&A]
Hello. I am trying to bind to the API of the [tdlib](https://github.com/tdlib/td) library. Their library has abstract classes as well as a regular one. Abstract classes are bound without any problems. The problem arises when binding final classes. Their library uses its smart pointer. But as I understand it, it can be replaced with a unique pointer.

It is difficult to link to a github with a class that I am binding to because their library generates this class automatically. I uploaded the generated file here: https://github.com/MarshalX/td_api

In general, this is what I am trying to do:
```cpp
#include <iostream>
#include <limits>
#include <map>
#include <memory>
#include <sstream>

#include <pybind11/pybind11.h>

#include <td/telegram/td_api.h>


namespace py = pybind11;
namespace td_api = td::td_api;


void ping() {
    py::print("pong");
}

PYBIND11_DECLARE_HOLDER_TYPE(Type, td_api::object_ptr<Type>)
PYBIND11_MODULE(tdlib_native, m) {
    m.def("ping", &ping);

    py::class_<td_api::LogStream,  td_api::object_ptr<td_api::LogStream>>(m, "LogStream");

    py::class_<td_api::logStreamDefault, td_api::object_ptr<td_api::logStreamDefault>>(m, "logStreamDefault");
}
```

LogStream binds without problems, but logStreamDefault does not (`Symbol not found: __ZTIN2td6td_api16logStreamDefaultE`). 

The CMake file and so on can be viewed in my repository https://github.com/MarshalX/tdlib

I would be grateful for any tip!

<details>
  <summary>LogStream class</summary>
  
<img width="601" alt="obraz" src="https://user-images.githubusercontent.com/15520314/128142922-cb16b9d9-261a-4078-98ab-9e5c610e620a.png">
  
</details>

<details>
  <summary>logStreamDefault class</summary>
  
<img width="920" alt="obraz" src="https://user-images.githubusercontent.com/15520314/128143162-0508cfa0-36e3-4000-a963-1321dc786111.png">

  
</details>

**Comment by MarshalX** [comment]
Iam also tried this one:
```cpp
py::class_<td::TlObject>TlObject(m, "TlObject");
py::class_<td_api::Object>Object(m, "Object", TlObject);

py::class_<td_api::LogStream>LogStream(m, "LogStream", Object);
py::class_<td_api::logStreamDefault>(m, "logStreamDefault", LogStream);
```

**Comment by henryiii** [comment]
We do not support std::unique_ptr yet, see the `smart_holder` branch where this is being developed - but normal smart pointers should be fine. But, anyway, to me it sounds like you are not linking to part of that library, since you get a symbol not found.

**Reply by MarshalX** [reply]
@henryiii  thanks your for your response! But if I not linking correct, does some code will works just from C++ part? I can use this class in m.def without any problems. Example:
```cpp
namespace py = pybind11;
namespace td_api = td::td_api;


void ping() {
    py::print("pong");

    auto test = td_api::logStreamDefault();
    auto id = test.get_id();

    py::print(id);
}

PYBIND11_MODULE(tdlib_native, m) {
    m.def("ping", &ping);
}
```
This code works and print id that stored in class field (I displayed it in the spoiler in the main message). 

Moreover, I bind an abstract class from the same file (td_api) without any problems ...

**Reply by MarshalX** [reply]
So I tried with pybind from smart_holder branch. There is nothing new( 

```cpp
#include <pybind11/smart_holder.h>

#include <td/telegram/td_api.h>


namespace py = pybind11;
namespace td_api = td::td_api;


void ping() {
    py::print("pong");
}

PYBIND11_DECLARE_HOLDER_TYPE(T, td_api::object_ptr<T>);
PYBIND11_TYPE_CASTER_BASE_HOLDER(td_api::logStreamDefault, td_api::object_ptr<td_api::logStreamDefault>)
PYBIND11_MODULE(tdlib_native, m) {
    m.def("ping", &ping);

    py::class_<td::TlObject, td_api::object_ptr<td::TlObject>>TlObject(m, "TlObject");
    py::class_<td_api::Object, td_api::object_ptr<td_api::Object>>Object(m, "Object", TlObject);

    py::class_<td_api::LogStream, td_api::object_ptr<td_api::LogStream>>LogStream(m, "LogStream");
    py::class_<td_api::logStreamDefault, td_api::object_ptr<td_api::logStreamDefault>>(m, "logStreamDefault", LogStream)
        .def(py::init<>())
        .def("get_id", &td_api::logStreamDefault::get_id);

}
``` 
if I will comment this lines
```cpp
 py::class_<td_api::logStreamDefault, td_api::object_ptr<td_api::logStreamDefault>>(m, "logStreamDefault", LogStream)
        .def(py::init<>())
        .def("get_id", &td_api::logStreamDefault::get_id);
```
it will working without any errors during the import

**Reply by MarshalX** [reply]
@henryiii also it works fine if I wrapped the class (I cant inheritance because class is final) 
```cpp
#include <pybind11/smart_holder.h>
#include <td/telegram/td_api.h>

namespace py = pybind11;
namespace td_api = td::td_api;

class logStreamDefault_ {
public:
  logStreamDefault_() {
    _instance = td_api::logStreamDefault();
  }

  std::int32_t get_id() {
    return _instance.get_id();
  }
private:
  td_api::logStreamDefault _instance;
};

PYBIND11_DECLARE_HOLDER_TYPE(T, td_api::object_ptr<T>);
PYBIND11_TYPE_CASTER_BASE_HOLDER(logStreamDefault_, td_api::object_ptr<logStreamDefault_>)
PYBIND11_MODULE(tdlib_native, m) {
    py::class_<logStreamDefault_, td_api::object_ptr<logStreamDefault_>>(m, "logStreamDefault")
        .def(py::init<>())
        .def("get_id", &logStreamDefault_::get_id);
}
```
